### PR TITLE
Convert trigger viewer to blade

### PR DESCRIPTION
**Pull request by Jamiras**
Changes:
* ID column numbers are muted and right-aligned
* Notes section is now collapsible, defaulting to collapsed
* Indirect pointer notes are supported.
* Only code notes needed for the display are fetched (instead of all code notes for a game)

image

### CODE REVIEW

**Comment by wescopeland** [comment]
I'm in the process of reviewing this but it'll likely be a day or two before I'm finished.

File: app/Platform/Services/TriggerDecoderService.php [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+                    $mem = substr($mem, 0, 2) . ' ' . substr($mem, 2);
+                    break;
+
+                default:
+                    $size = $mem[2];
+                    break;
+            }
+
+            $mem = substr($mem, 3);
+        } elseif ($end > 2 && $mem[0] === 'f' || $mem[0] === 'F') {
```
Comment (wescopeland) [inline app/Platform/Services/TriggerDecoderService.php:65]: There may be a subtle bug here. This is evaluated as:

```php
// ((A && B) || C)
(($end > 1 && $mem[0] === 'f') || $mem[0] === 'F') {
```

However, I believe the intent is for this to be evaluated as `(A && (B || C))`.

This also occurs with the subsequent `elseif` comparing `$mem[0]` to "h" or "H".

File: app/Platform/Services/TriggerDecoderService.php [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+            'SourceAddress' => $lMemory,
+            'Operator' => $cmp,
+            'TargetType' => $rType,
+            'TargetSize' => $rSize,
+            'TargetAddress' => $rMemVal,
+            'HitTarget' => $hits,
+        ];
+    }
+
+    public function decode(string $serializedTrigger): array
```
Comment (wescopeland) [inline app/Platform/Services/TriggerDecoderService.php:280]: `->map()` should eliminate the need for writing nested `for` loops:

```php
public function decode(string $serializedTrigger): array
{
    $serializedGroups = collect(preg_split("/(?<!0x)[S$]/", $serializedTrigger));

    return $serializedGroups->map(function ($groupString, $index) {
        $reqs = explode('_', $groupString);
        $isIndirect = false;

        $conditions = collect($reqs)->filter()->map(function ($req) use (&$isIndirect) {
            $condition = $this->parseCondition($req);
            $condition['IsIndirect'] = $isIndirect;

            if ($condition['SourceType'] === 'Value') {
                $condition['SourceTooltip'] = hexdec($condition['SourceAddress']);
            }
            if ($condition['TargetType'] === 'Value') {
                $condition['TargetTooltip'] = hexdec($condition['TargetAddress']);
            }

            $isIndirect = $condition['Flag'] === 'Add Address';
            return $condition;
        });

        return [
            'Label' => ($index === 0) ? 'Core Group' : "Alt Group $index",
            'Conditions' => $conditions->all()
        ];
    })->all();
}
```

File: resources/views/pages-legacy/achievementInfo.blade.php [hunk 3]
```diff
@@ -396,7 +396,15 @@ function ResetProgress() {
 
         echo "<code>" . htmlspecialchars($achMem) . "</code>";
         echo "<li>Mem explained:</li>";
-        echo "<code>" . getAchievementPatchReadableHTML($achMem, $codeNotes) . "</code>";
+
+        $service = new TriggerDecoderService();
```
[the remark this answers is not in the document]
Comment (Jamiras) [reply resources/views/pages-legacy/achievementInfo.blade.php:400]: done

File: resources/views/pages-legacy/leaderboardinfo.blade.php [hunk 4]
```diff
@@ -157,11 +159,37 @@
                 }
             }
 
-            getCodeNotes($gameID, $codeNotes);
-            ExplainLeaderboardTrigger('Start', $memStart, $codeNotes);
-            ExplainLeaderboardTrigger('Cancel', $memCancel, $codeNotes);
-            ExplainLeaderboardTrigger('Submit', $memSubmit, $codeNotes);
-            ExplainLeaderboardTrigger('Value', $memValue, $codeNotes);
+            $explainLeaderboardTrigger = function(string $name, string $triggerDef) use ($gameID) : void
```
[the remark this answers is not in the document]
Comment (Jamiras) [reply resources/views/pages-legacy/leaderboardinfo.blade.php:162]: done

File: app/Platform/Services/TriggerDecoderService.php (hunk 2 above)
Comment (wescopeland) [reply app/Platform/Services/TriggerDecoderService.php:280]: The only thing I'm shooting for is readability and maintainability. I agree with your points and believe the inner `foreach` is a good change.

**Review by wescopeland — APPROVED** [review]

**Pull request merged**